bson/primitive/decimal_test.go

```diff
@@ -144,3 +148,61 @@ func TestParseDecimal128(t *testing.T) {
 		}
 	}
 }
+
+func TestDecimal128_JSON(t *testing.T) {
+	t.Run("roundTrip", func(t *testing.T) {
+		decimal := NewDecimal128(0x3040000000000000, 12345)
+		bytes, err := json.Marshal(decimal)
+		assert.Nil(t, err, "json.Marshal error: %v", err)
+		got := NewDecimal128(0, 0)
+		err = json.Unmarshal([]byte(bytes), &got)
+		assert.Nil(t, err, "json.Unmarshal error: %v", err)
+		assert.Equal(t, decimal.h, got.h, "expected h: %v got: %v", decimal.h, got.h)
+		assert.Equal(t, decimal.l, got.l, "expected l: %v got: %v", decimal.l, got.l)
+	})
+	t.Run("unmarshal extendedJSON", func(t *testing.T) {
+		want := NewDecimal128(0x3040000000000000, 12345)
+		extJSON := fmt.Sprintf(`{"$numberDecimal": %q}`, want.String())
+
+		got := NewDecimal128(0, 0)
+		err := json.Unmarshal([]byte(extJSON), &got)
+		assert.Nil(t, err, "json.Unmarshal error: %v", err)
+		assert.Equal(t, want.h, got.h, "expected h: %v got: %v", want.h, got.h)
+		assert.Equal(t, want.l, got.l, "expected l: %v got: %v", want.l, got.l)
+	})
+	t.Run("unmarshal null", func(t *testing.T) {
+		want := NewDecimal128(0, 0)
+		extJSON := `null`
+
+		got := NewDecimal128(0, 0)
+		err := json.Unmarshal([]byte(extJSON), &got)
+		assert.Nil(t, err, "json.Unmarshal error: %v", err)
+		assert.Equal(t, want.h, got.h, "expected h: %v got: %v", want.h, got.h)
+		assert.Equal(t, want.l, got.l, "expected l: %v got: %v", want.l, got.l)
+	})
+	t.Run("unmarshal", func(t *testing.T) {
+		cases := append(bigIntTestCases,
+			[]bigIntTestCase{
+				{s: "-0001231.453454000000565600000000E-21", h: 0xafe6000003faa269, l: 0x81cfeceaabdb1800},
+				{s: "12345E+21", h: 0x306a000000000000, l: 12345},
+				{s: "0.10000000000000000000000000000000000000000001", remark: "parse fail"},
+				{s: ".125e1", h: 0x303c000000000000, l: 125},
+				{s: ".125", h: 0x303a000000000000, l: 125},
+			}...)
+		for _, c := range cases {
+			input := fmt.Sprintf(`{"foo": %q}`, c.s)
+			var got map[string]Decimal128
+			err := json.Unmarshal([]byte(input), &got)
+
+			switch c.remark {
+			case "overflow", "parse fail":
+				assert.NotNil(t, err, "expected Unmarshal error, got nil")
+			case "", "rounding", "subnormal", "clamped", "NaN", "Infinity", "-Infinity":
```

Comment:
nit: Instead of listing out all of the possible values here, should we just make this `default`?